edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs

```diff
@@ -19,30 +20,68 @@ public class KubernetesRuntimeInfoProvider : IRuntimeInfoProvider, IRuntimeInfoS
     {
         readonly string deviceNamespace;
         readonly IKubernetes client;
-        readonly ConcurrentDictionary<string, ModuleRuntimeInfo> moduleRuntimeInfo;
+        readonly ConcurrentDictionary<string, ImmutableList<KeyValuePair<string, ModuleRuntimeInfo>>> moduleRuntimeInfo;
         readonly IModuleManager moduleManager;
 
         public KubernetesRuntimeInfoProvider(string deviceNamespace, IKubernetes client, IModuleManager moduleManager)
         {
             this.deviceNamespace = Preconditions.CheckNonWhiteSpace(deviceNamespace, nameof(deviceNamespace));
             this.client = Preconditions.CheckNotNull(client, nameof(client));
             this.moduleManager = Preconditions.CheckNotNull(moduleManager, nameof(moduleManager));
-            this.moduleRuntimeInfo = new ConcurrentDictionary<string, ModuleRuntimeInfo>();
+            this.moduleRuntimeInfo = new ConcurrentDictionary<string, ImmutableList<KeyValuePair<string, ModuleRuntimeInfo>>>();
         }
 
-        public void CreateOrUpdateAddPodInfo(string podName, V1Pod pod)
+        public void CreateOrUpdateAddPodInfo(V1Pod pod)
         {
-            ModuleRuntimeInfo runtimeInfo = pod.ConvertToRuntime(podName);
-            this.moduleRuntimeInfo.AddOrUpdate(podName, runtimeInfo, (_, existing) => runtimeInfo);
+            string podName = pod.Metadata.Name;
+            string moduleName = pod.Metadata.Labels[Constants.K8sEdgeModuleLabel];
+            var newPair = new KeyValuePair<string, ModuleRuntimeInfo>(podName, pod.ConvertToRuntime(moduleName));
+
+            this.moduleRuntimeInfo.AddOrUpdate(
+                moduleName,
+                ImmutableList.Create(newPair),
+                (_, existing) =>
+                {
+                    return existing.FirstOption(pair => pair.Key == podName)
+                        .Map(pair => existing.Replace(pair, newPair))
+                        .GetOrElse(() => existing.Add(newPair));
+                });
         }
 
-        public bool RemovePodInfo(string podName)
+        public bool RemovePodInfo(V1Pod pod)
         {
-            return this.moduleRuntimeInfo.TryRemove(podName, out _);
+            var podName = pod.Metadata.Name;
+            var moduleName = pod.Metadata.Labels[Constants.K8sEdgeModuleLabel];
+            if (this.moduleRuntimeInfo.TryGetValue(moduleName, out var existing))
+            {
+                Option<ImmutableList<KeyValuePair<string, ModuleRuntimeInfo>>> updatedList = existing.FirstOption(pair => pair.Key == podName)
```

Comment:
Same here. Lets inline `podName`
```suggestion
                Option<ImmutableList<KeyValuePair<string, ModuleRuntimeInfo>>> updatedList = existing.FirstOption(pair => pair.Key == pod.Metadata.Name)
```